### ISSUE TITLE
Remote Containers breaks SSH-forwarded GPG agent with `updatestartuptty`

### ISSUE THREAD

**Issue by jordemort** [microsoft/vscode-remote-release#6566]
I run VSCode on a MacBook Pro, and develop remotely on an Ubuntu server, using VSCode Remote SSH.

The GPG key I use for signing Git commits is on a YubiKey. I have GPG  on macOS configured to talk to the YubiKey, and then I forward the gpg-agent socket from macOS to my Linux machine over SSH, using a line like this in my `ssh_config`:

```
# 2001 is my uid on the remote machine
RemoteForward /run/user/2001/gnupg/S.gpg-agent /Users/jordan/.gnupg/S.gpg-agent.extra
```

This has been working well so far with just Remote SSH, but I recently decided to add a devcontainer into the mix. Everything works well, except that the setup that Remote Containers attempts to do before starting the container breaks gpg-agent in a way that I am unable to sign commits in the container.

At first, I thought this was a simple matter of needing to have the forwarded socket bound into the container, so I switched to using a `docker-compose.yml` file for my devcontainer with `/run/user` bound in:

```yaml
services:
  devcontainer:
    ...
    volumes:
      - /run/user:/run/user
      ...
```

However, attempting to sign anything in the container still fails:

```
$ echo "foo" | gpg --sign > /dev/null
gpg: WARNING: unsafe permissions on homedir '/home/builder/.gnupg'
gpg: can't connect to the agent: End of file
gpg: can't connect to the agent: IPC connect call failed
gpg: keydb_search failed: No agent running
gpg: no default secret key: No agent running
gpg: signing failed: No agent running
```

After doing some debugging, I believe this is because Remote Containers is attempting to run `gpg-connect-agent updatestartuptty` before starting the container. Because the gpg-agent is being forwarded from a remote host, `updatestartuptty` is meaningless and invalid in this context. This seems to somehow break the connection between the gpg-agent process on my Mac and the socket forwarded to my Ubuntu host; any further attempts to sign things are unsuccessful, either inside or outside the container.

If I then open a terminal window and start another SSH session to my Ubuntu host, it re-creates the forwarded gpg-agent socket, and then I am able to sign commits again. So I am able to work around it, but I need to open up another terminal window after I open my project in the devcontainer and keep it running as long as I am working, which is somewhat less than idea.

I'm not sure what the right answer here is, because `updatestartuptty` is probably the right thing to do for a local agent, and there's no easy programmatic way that I can think of to distinguish a forwarded gpg-agent from a local one. I think maybe an additional setting is needed to tell Remote Containers not to attempt `updatestartuptty`, but it would not make sense as a devcontainer.json or per-project setting, since someone else without a fowarded gpg-agent might want to work on the same project.

- VSCode Version: 1.66.0 (Universal) Commit: e18005f0f1b33c29e81d732535d8c0e47
- Local OS Version: macOS Montery 12.3 (21E230)
- Remote OS Version: Ubuntu 20.04, Docker 20.10.14
- Remote Extension/Connection Type: SSH and Docker
- Logs: (gathered from gpg-agent by adding `log-file /tmp/gpg-agent.log` to `~/.gnupg/gpg-agent.conf`)

```
2022-04-05 11:27:33 gpg-agent[76214] command 'UPDATESTARTUPTTY' failed: Forbidden
2022-04-05 11:27:33 gpg-agent[76214] DBG: chan_8 -> ERR 67109115 Forbidden <GPG Agent>
2022-04-05 11:27:33 gpg-agent[76214] DBG: chan_8 <- [eof]
```

Steps to Reproduce:

1. Configure SSH to forward your gpg-agent socket, as above.
2. Configure a project with a devcontainer to bind /run/user, as above.
3. Open the project in the devcontainer on your SSH remote.
4. Observe that gpg-agent will no longer communicate with you after having received an invalid `updatestartuptty` command.

Does this issue occur when you try this locally?: No
Does this issue occur when you try this locally and all extensions are disabled?: No

**Comment by jordemort** [comment]
Managed to capture a log from Remote Containers confirming that it's doing the bad thing:

```
[10107 ms] Start: Run in Host: gpg-connect-agent updatestartuptty /bye
```

**Comment by jeremyn** [comment]
I have a slightly different problem: I have a GPG key installed directly on a remote Ubuntu 20.04 host, not forwarded from a local system, but I can't sign stuff with that key inside a container. I can see the key inside the container with `gpg --list-secret-keys`, but get an error like `No such file or directory <Pinentry>` with for example `echo "secret" | gpg --debug-level 9 --clearsign`.

@jordemort, maybe those commands will help you get extra info, or you can try installing the key directly and see if you have better luck than I did. Also, the message `gpg: WARNING: unsafe permissions on homedir '/home/builder/.gnupg'` suggests you've manually created the `.gnupg` directory in the container, instead of letting VS Code create it. The devs might ask you about that.

I suspect using GPG from a remote-SSH host inside a container just doesn't work. I'd be interested if anyone could successfully sign anything with GPG with a minimal environment like:

`Dockerfile`:

```dockerfile
FROM ubuntu:20.04

RUN apt-get update && DEBIAN_FRONTEND=noninteractive apt-get install -y --no-install-recommends \
  gpg \
  && rm -rf /var/lib/apt/lists/*
```

`.devcontainer/devcontainer.json`:
```json
{
  "dockerFile": "../Dockerfile"
}
```

Also, though @jordemort has worked around the problem by signing the code outside of the container, I suppose because he's making it available in the container with a bind mount, this is not a good workaround if your code is in a volume, unless you want to go digging around in `/var/lib/docker/volumes` on the host.

**Comment by jordemort** [comment]
> you can try installing the key directly and see if you have better luck than I did

I can not, because the key was generated on the YubiKey, and it is impossible to extract it. Being able to do so would defeat the point of the YubiKey.

> Also, the message gpg: WARNING: unsafe permissions on homedir '/home/builder/.gnupg' suggests you've manually created the .gnupg directory in the container, instead of letting VS Code create it. The devs might ask you about that.

I haven't touched it; VSCode appears to have made a copy of what was in my home directory there.

> Also, though @jordemort has worked around the problem by signing the code outside of the container

I am signing the code inside the container. I have to start an additional SSH session to re-establish the socket forwarding, but once the socket is recreated, singing commits work both inside and outside of the container, at least until the next time VSCode restarts the container. Before it is re-established, signing fails both inside and outside of the container; the `UPDATESTARTUPTTY` seems to gum things up to the point that once it's issued, the agent socket is no longer usable until it is recreated.

**Comment by jeremyn** [comment]
@jordemort:

> I can not, because the key was generated on the YubiKey, and it is impossible to extract it. Being able to do so would defeat the point of the YubiKey.

That's fair, though you could create a temporary GPG key to test with.

> I haven't touched it; VSCode appears to have made a copy of what was in my home directory there.

I don't think VS Code is supposed to do that. See https://github.com/microsoft/vscode-remote-release/issues/5644#issuecomment-1031525648 from two months ago where @chrmarti wrote "Make sure the container's `~/.gnupg` folder is empty, we don't overwrite any files in there and that might explain why you don't see any keys."

> I am signing the code inside the container. ...

In my case I've signed stuff outside the container to unlock the key in `gpg-agent` before trying, and failing, to sign stuff inside the container. So our situations should be similar enough, but we're getting different results. When you're getting the error signing inside the container, can you see the keys inside the container with `gpg --list-secret-keys`?

**Comment by jordemort** [comment]
> That's fair, though you could create a temporary GPG key to test with.

I could, but doing so would be irrelevant to my problem. My issue is specifically with regard to a key that is only available via a forwarded gpg-agent.

> I don't think VS Code is supposed to do that...

It appears currently that it's a problem with the setup script of my dotfiles repository, and I just need to chmod that directory to 0700. That makes the "unsafe permissions" message go away, but does not otherwise change the behavior I am describing here. I also had the same problem before I configured my dotfiles repository. Again, I don't think this is very relevant, because once I recreate the socket, singing commits works, regardless of the permissions on .gnupg.

>  So our situations should be similar enough, but we're getting different results. 

I am reasonably certain that the cause of my issue is Remote Containers issuing `UPDATESTARTUPTTY`; otherwise, my setup works fine. It has been working fine with Remote SSH alone for a while; it is only when I add Remote Containers into the mix that I have the problem. 

I am unsure if your problem is related to mine; in your case, it sounds like gpg-agent is trying to start inside of your container, instead of communicating with the agent (and using the pinentry) outside of the container. Perhaps you should try binding `/run/user` into your container to see if you get better results; either that, or make the gpg installation in your container is complete and there is a pinetry program available; since the source is not available, I couldn't tell you if VSCode does anything to handle forwarding a gpg agent socket on its own or not; if it does, I wasn't able to get it to work, which is why I resorted to the bind mount.

When the agent socket is broken, `gpg --list-secret-keys` says this:

```
gpg: can't connect to the agent: End of file
```

**Comment by jeremyn** [comment]
The issue I linked in my previous comment suggests that when GPG is working in remote containers (1) VS Code copies some but not all `.gnupg` files from the host into the container and (2) gpg-agent should not be running in the container. Not sure if that matters for either of us, but, just to be clear.

I think in my case gpg is connecting to the host agent from within the container because the debug output has `gpg: DBG: connection to agent established` but then later has `Forbidden <GPG Agent>` when trying to set options, which I think is typical when GPG tries to pass commands back to a forwarded agent. If this was all internal to the container it would say something like "key missing" instead of "forbidden".

Another related issue for me is "Support command-line pinentry for GPG" https://github.com/microsoft/vscode-remote-release/issues/5730. It also discusses tty/updatestartuptty stuff, and in https://github.com/microsoft/vscode-remote-release/issues/5730#issuecomment-946497943 @chrmarti repeatedly refers to some sort of "shell interference" which may (?) be related to the behavior you're seeing. Unfortunately that issue was closed when the reporter was able to use a graphical GUI, since they were using remote containers directly on a Mac without SSH.

Anyway I don't mean to take over your issue. I'm surprised you got it to work at all, even partly. I look forward to seeing how this goes, at least to get some confirmation on what configurations are supported. @chrmarti (or whoever), if you think my problem is worth investigating separately, please let me know and I can make a new issue.

**Comment by jeremyn** [comment]
I looked into this some more:

I can confirm the `Run in Host: gpg-connect-agent updatestartuptty /bye` message @jordemort saw, under "Remote-Containers: Show Container Log" in the command palette.

I also have a YubiKey so I tried this (separately) with both the GPG key installed directly and with the YubiKey. My "remote" Ubuntu 20.04 server environment is a VM in VMware, so I can just pass the YubiKey into the VM as a device.

@jordemort In my case, I just need to run `gpg-connect-agent updatestartuptty /bye` in a separate non-VS Code SSH session and then I can use GPG in the container with the YubiKey. Also, I have a GPG PIN on my YubiKey, so both the directly installed key and the YubiKey require pinentry. The `updatestartuptty` command lets pinentry show up on the non-VS Code session in both cases, unfortunately actually typing in a GPG passphrase-or-PIN is usually garbled as in [these](https://unix.stackexchange.com/q/630101) [issues](https://unix.stackexchange.com/q/325021) on Stack Exchange. Also unfortunately, with a directly installed key, I can't unlock the key in the non-VS Code session and then use that in the container -- it needs to be unlocked again when the container tries to use the key. Good news though, I can unlock the YubiKey in the non-VS Code session and then use it in the container.

So I think we're seeing basically the same thing. In your case I agree that adding some setting to stop VS Code from running `updatestartuptty` when starting the container is a fix. You start with a forwarding chain of Mac < remote Ubuntu. VS Code adds itself to the chain as Mac < remote Ubuntu < container and runs `updatestartuptty` on remote Ubuntu, breaking the chain. Your new session fixes the chain.

In my case of just remote Ubuntu < container, there is no good fix. The key has to be entered on the remote host, but there might not be any session open there. This could be clarified in the VS Code documentation at least.

(I use a Windows desktop and I'd like to set up forwarding like your Mac, unfortunately Windows doesn't have a great story for GPG agent forwarding, particularly with a YubiKey.)

**Comment by chrmarti** [comment]
To clarify what Remote-Containers does:
- Check if GPG forwarding should be between the container and the ssh server or the container and the local machine:
  - Run `gpgconf --list-dir agent-extra-socket` on the ssh server first and if that returns a socket, forward to that. Otherwise rerun this command on the local machine and forward to the socket returned there.
- Copy pubring.kbx, pubring.gpg and trustdb.gpg to the container. That appears to be needed for the forwarding to work.
- Run `gpg-connect-agent updatestartuptty /bye` to make sure the gpg-agent is running.
  - `updatestartuptty` seemed to avoid some IPC error on Mac, not sure that is always correct. It seems to update the gpg-agent with the current `GPG_TTY` and `DISPLAY`, if I read the documentation correctly.

**Comment by jeremyn** [comment]
@chrmarti Recent (since 2016) versions of GPG [always create](https://wiki.gnupg.org/AgentForwarding) the extra socket, so your `gpgconf` test will always return true for the "ssh server", which I guess means the remote host, if that system is at all modern. So that is not a good test. I'm unclear what you mean by rerunning the command on the local machine.

`gpg-connect-agent updatestartuptty /bye` seems too aggressive if you just want to check that gpg-agent is running, since it forcibly redirects pinentry requests to wherever it runs. You should assume the user is already directing requests to their preferred terminal and not interfere with that. I don't know about the IPC error on Mac, though.

**Comment by chrmarti** [comment]
The lookup for the extra socket is mainly to make sure GPG is installed.

I will look into the `updatestartuptty` issue.

**Comment by chrmarti** [comment]
To clarify: We only use the extra socket, if that is not configured on the SSH server we fall back to the extra socket on the local machine and if that is not configured we do not forward the GPG agent.

**Comment by chrmarti** [comment]
The Remote-Containers extension is forwarding the local extra socket to the container. It also forwards the ssh-agent socket and the Git credential manager (not by socket, but by using the Git CLI).

For security we set up "workspace trust" such that you first need to trust the workspace folder with the devcontainer.json before you can reopen it in a container. Do you see any other problems with this?

**Comment by jeremyn** [comment]
@chrmarti I hope this doesn't sound argumentative, but your response seems to miss that we're doing a second hop through an intermediate server. I'm still not entirely sure we're talking about the same thing. Have you personally tried this or inspected the internal Remote-Containers code or are you answering based on the public documentation such as "[Sharing GPG Keys](https://code.visualstudio.com/docs/remote/containers#_sharing-gpg-keys)"? The documentation there doesn't mention using Remote-SSH with Remote-Containers.

You can see in my comment at https://github.com/microsoft/vscode-remote-release/issues/6608#issuecomment-1100419568 what I tried and what did and didn't happen. In that case on my local system I was using the GPG installed with Git Bash, not Gpg4Win. The key data got copied over but the container couldn't see any secret keys so any socket forwarding it tried to do, failed. Also, looking at the "Sharing GPG Keys" link, the closest matching situation is for WSL which requires installing the `socat` package, which I don't think I had on the SSH server. I'm not super interested in digging into it more, guessing on packages to install etc, unless you can confirm what works for you.

I vaguely understand the VS Code workspace trust concept, but you can see from my comment in the other issue that I consider Remote-SSH with Remote-Containers as a sort of sandbox. So in my case the answer would be that I trust the code and the dependencies to run in the contained environment specifically because they are limited in accessing anything outside of it.

**Comment by jordemort** [comment]
Can we spin up some other issues to discuss the the broader aspects of Remote SSH + Remote Containers, or should I rename this one and open up another issue specifically regarding the use of `updatestartuptty`? My one and only request in this issue is to be able to make Remote-Containers stop doing an `updatestartuptty`.

**Comment by jeremyn** [comment]
@jordemort I'm happy to organize the issues however is best, as I wrote in an earlier comment. 

However it looks like the meta-problem is that using Remote-SSH and Remote-Containers together is new and they haven't really locked down all the permutations yet, and in particular all the assumptions about the GPG pinentry terminal exclude an intermediate SSH server. Sorting that out includes understanding the basic flow of socket forwarding etc in the various cases that we've been discussing. Adding an option `disableUpdateStartupTTY` would be a hack targeted to this specific GitHub issue. That's probably not worth it especially since you can work around the problem just by opening another SSH session.

What do you think?

**Comment by chrmarti** [comment]
@jeremyn Opened https://github.com/microsoft/vscode-remote-release/issues/6667 to continue the discussion.

**Comment by jordemort** [comment]
> We only use the extra socket

I think this is the piece I was missing. I was forwarding the extra socket from my Mac to my Linux host, but I was forwarding it to the regular socket on the Linux host. That is to say, I was forwarding Mac `S.gpg-agent.extra` to Linux `S.gpg-agent`. So there was no functional extra socket on the Linux host.

I changed my configuration to forward `S.gpg-agent.extra` to both `S.gpg-agent` and `S.gpg-agent.extra` on Linux, and now things just work without any additional hacks - I no longer have to bind `/run/user` or do any special setup in the devcontainer. I can now open up devcontainers and create signed commits without any additional hacks or reconnections.

For posterity, here is the line I originally had in my `ssh_config`:

```
  RemoteForward /run/user/2001/gnupg/S.gpg-agent /Users/jordan/.gnupg/S.gpg-agent.extra
```

...and here is the line I added that made things start working:

```
  RemoteForward /run/user/2001/gnupg/S.gpg-agent.extra /Users/jordan/.gnupg/S.gpg-agent.extra
```

Since this now has the "Works on my machine" seal of approval, I'll close this.

**Comment by chrmarti** [comment]
@jordemort Would omitting `updatestartuptty` still be an improvement?

**Comment by jordemort** [comment]
@chrmarti I've been using it as a daily driver for over a week now with no further problems, now that I'm forwarding the extra socket. I think the root cause of whatever was causing updatestartuptty to cause problems was corrected by this; my working theory is that something like this was happening:

- systemd user unit on SSH remote is a socket activated service that listens on both S.gpg-agent and S.gpg-agent.extra
- When I connected, my original ssh config overwrote and forwarded S.gpg-agent but not S.gpg-agent.extra
- VSCode sees that S.gpg-agent.extra exists (because the systemd user unit created it) and tries to connect to it
- This caused systemd to start gpg-agent on the SSH remote, which may either be overwriting S.gpg-agent again or just coming up as an empty agent with no keys
- 💀 Bad things happen, as a result of S.gpg-agent and S.gpg-agent.extra being owned by two different gpg-agent processes running on different machines 💀

Now that I am forwarding to both S.gpg-agent and S.gpg-agent.extra, systemd is fully out of the loop no matter which socket is used, and when VSCode connects to S.gpg-agent.extra, it is talking to my forwarded agent, which is what I want.

I feel at this point that updatestartuptty breaking everything was probably more a symptom of attempting to talk to an agent when things where in such a confused state, rather than a result of updatestartuptty itself.

However, I'm still not sure it's correct for Remote Containers to be issuing it. As I understand it, updatestartuptty tells gpg-agent that you want to change the display that the pinentry program shows up on, re: [gpg documentation](https://www.gnupg.org/documentation/manuals/gnupg/Agent-UPDATESTARTUPTTY.html) - it doesn't seem to me like this is the right thing to do, because if a user has configured a custom pinentry program, they would also have to ensure that that pinentry program is available and usable inside the container, which is particularly unlikely if you're using something like pinentry-mac. On the other hand it seems like the extra socket treats pinentry differently, re: [other gpg documentation](https://wiki.gnupg.org/AgentForwarding). In my case it doesn't seem to break anything, but in my case gpg-agent is also serving as my SSH agent (and my SSH key is actually the same key as my GPG key) so my pinentry gets invoked to unlock the key before the SSH connection is even established; singing for me after the key is unlocked involves physically pressing the button on the key and the pinentry is not invoked, so between the particulars of my setup and GPG's different handling of pinetry on the extra socket, I can't tell if updatestartuptty is actually doing anything bad at this point or not.

**Comment by jordemort** [comment]
> At best it doesn't break anything, at worst it steals focus from the user's preferred terminal.

I think that's right. I would be curious if there was any reasoning attached to the commit that added updatestartuptty to Remote Containers - is it there because it fixed something else that was broken? Would anyone else's setup break, if Remote Containers no longer did updatestartuptty?

> I think you always need gpg-agent running for GPG to access a socket, so having it run once on the local machine and once on the Remote-SSH machine is unavoidable.

No, this is wrong. You only want it running once. Running it anywhere other than where it has access to the key is useless. In my case the problem was that my ssh_config only forwarded the main socket, and not the extra socket. VSCode only communicates with the extra socket, so when tried talking to that, it caused systemd to to incorrectly start an additional agent on the SSH remote. Now that I am also forwarding the extra socket, VSCode reaches the agent running on my local machine. There is no gpg-agent running on the SSH remote, or inside the container; when things are working properly, there is only one agent running on my local machine.

> Having to forward from the local machine to two different sockets on the Remote-SSH machine seems inelegant. It would be better if you could instruct Remote-Containers to forward from the S.gpg-agent socket on the Remote-SSH machine into the container.

I am happy with the current state of things. S.gpg-agent.extra is intended to be fowarded; S.gpg-agent is not. It makes sense that Remote Containers would prefer S.gpg-agent.extra in this case, and it would be possibly dangerous to make it use S.gpg-agent instead. Now that I know, it is not an ongoing source of inconvenience, it's just another line in my ssh_config. I suspect that the subset of users that want to use Remote SSH with Remote Containers and that also want to sign their commits using a forwarded gpg-agent is pretty small, and anyone who is forwarding their gpg-agent is already familiar with fighting with their ssh_config.

**Comment by jeremyn** [comment]
@jordemort 

> I would be curious if there was any reasoning attached to the commit that added updatestartuptty to Remote Containers

@chrmarti wrote at https://github.com/microsoft/vscode-remote-release/issues/6566#issuecomment-1094642903 that "`updatestartuptty` seemed to avoid some IPC error on Mac", so I assume it's something that was added before Remote-SSH with Remote-Containers was a thing.

> No, this is wrong. You only want it running once.

I'm not interested in a super deep dive here but gpg requires the agent, look at the note for `--use-agent` in the [gpg man page](https://linux.die.net/man/1/gpg). Also in @chrmarti's quote he also says they run `gpg-connect-agent` specifically to "make sure the gpg-agent is running" (on the Remote-SSH machine). A socket is just a file on the system, sitting there. gpg-agent is the thing that connects a GPG request (and in your case, an SSH request), to that socket file. Otherwise I don't know how it would work.

> S.gpg-agent.extra is intended to be fowarded; S.gpg-agent is not. It makes sense that Remote Containers would prefer S.gpg-agent.extra in this case, and it would be possibly dangerous to make it use S.gpg-agent instead.

So again I'm not trying to get super into here but the usual pattern as I understand it is: local S.gpg-agent.extra > remote1 S.gpg-agent, remote1 S.gpg-agent > remote2 S.gpg-agent and so on. You can see this pattern explicitly at for example the well-known @drduh Yubikey guide at https://github.com/drduh/YubiKey-Guide#chained-gpg-agent-forwarding, as one example, but I don't think you'll find many double-forwarding examples in the very long history of gpg. I can think of good reasons for this but we don't need to get into it. The reason VS Code behaves the way it does is because of its, pardon me, lazy behavior of only checking for extra sockets, I assume because they only recently combined Remote-SSH and Remote-Containers and haven't worked it all out yet.

I don't have a detailed recommendation, but this gets back to my comment at https://github.com/microsoft/vscode-remote-release/issues/6667#issuecomment-1117462595 where I basically said VS Code Remote's GPG integration is too complex to explain responsibility and VS Code's current poorly-documented and opt-out behavior is a security issue.

**Comment by jordemort** [comment]
> I'm not interested in a super deep dive here but gpg requires the agent

It does not require that the agent be running on the local machine. I can prove using `ps` that in my setup where I'm signing commits in Remote Containers + Remote SSH that there is no agent process running in the container or on the ssh remote. The only place the agent runs is on my local machine. Any connections to S.gpg-agent or S.gpg-agent.extra on the remotes are forwarded to S.gpg-agent.extra on my local machine, through magic of SSH forwarding. This is what the `RemoteForward` lines in my ssh_config do.

> So again I'm not trying to get super into here but the usual pattern as I understand it is: local S.gpg-agent.extra > remote1 S.gpg-agent, remote1 S.gpg-agent > remote2 S.gpg-agent and so on. 

That is, indeed, the usual setup, and the one I started with. When forwarding the agent, the gpg documentation says to forward `S.gpg-agent.extra`. As far as I can tell, Remote Containers is attempting to conform to this pattern; it, too, is attempting to forward `S.gpg-agent.extra`. If I was using Remote Containers without Remote SSH, the way it behaves now would definitely be the right thing, and knowing how it works, it's easy for me to modify my setup so that it works for me too. If it decided to forward `S.gpg-agent` instead, that would be the right thing for me, but only because my `S.gpg-agent` is actually forwarded to an `S.gpg-agent.extra` - if there was a real agent on `S.gpg-agent`, then it would be dangerous, because that socket is not meant to be forwarded. I think it's better to fail for people like me in a not-dangerous way, that is easily corrected (maybe a little extra documentation about how it works would be good 😄) than for it to do something potentially dangerous.

> I don't have a detailed recommendation

None is required.

**Comment by jeremyn** [comment]
@jordemort

I admit I haven't tested this specifically myself, but if we were talking about pure GPG and SSH then what you're seeing would not be possible. A user manually moving from one server to another in a chain would have GPG/SSH support, and therefore `gpg-agent`, available at each step in the chain. However from @chrmarti's comment at https://github.com/microsoft/vscode-remote-release/issues/6566#issuecomment-1109612162 it does sound like VS Code Remote has some internal magic that supports "direct" forwarding from the local to the container, somehow bypassing the Remote-SSH machine. Maybe that's what you're seeing.

I don't know about which socket is "meant" to be forwarded but it is the case that, I think, the large majority of documentation and snippets support forwarding `S.gpg-agent` to `S.gpg-agent` for all hops after the first. If VS Code Remote intentionally expects different behavior, then I agree that should be documented, if possible.

**Comment by jordemort** [comment]
> if we were talking about pure GPG and SSH then what you're seeing would not be possible

@jeremyn I am sorry, but you are just flat-out wrong about how this works. Please think about it: my key lives on a YubiKey that is directly connected to my Mac. What could a gpg-agent process running on my Ubuntu SSH remote possibly do with that key? It has no way to access it - gpg-agent doesn't know how to talk to another gpg-agent. I'm not sharing the USB device over the network. I can't export the key to a file - the YubiKey only offers the opportunity to do that once, when the key is generated, and I turned it down - my key has never existed as a file and cannot exist as a file. There is no way for a remote agent to use my key, and yet I sign commits remotely all the time. This is because there is _no remote agent_ - everything is forwarded over SSH to my local agent.

All that is going on is here SSH forwarding, both of the SSH agent (`ForwardAgent yes` - ssh has built-in special handling of this case) and of the GPG agent socket (the `RemoteForward` lines - in this case ssh is just forwarding a UNIX socket to a UNIX socket and does not know anything else about it.) It works with or without VSCode in the loop; I can SSH into a remote server in my terminal and sign commits just fine, for example, without VSCode even running - I can even do the same thing in a container, if I set up the appropriate bind mounts by hand. Where VSCode is getting involved in this process is at the Remote Containers step, where it is either binding or forwarding the extra socket into the container automatically -- and that works just fine too, as long as the right thing is listening on the extra socket.

I suggest you read [this wiki page](https://wiki.gnupg.org/AgentForwarding) and the [ssh_config man page](https://man.openbsd.org/ssh_config) if you would like to get a better understanding of how this all works. Here is [where the magic happens](https://github.com/jordemort/dotfiles/blob/main/home/.ssh/config.d/default#L19-L21=) in my dotfiles.

**Comment by jeremyn** [comment]
@jordemort Again I'm not sure how deep I want to go here but I have seen that AgentForwarding wiki page, in fact I linked it a month ago in this very issue, and your SSH config looks exactly as I would expect.

Again it is important to understand that:

* gpg requires gpg-agent (from the gpg [man page](https://linux.die.net/man/1/gpg) under `--use-agent`); and
* gpg-agent functions as a "drop-in replacement" for ssh-agent if you use it for SSH authentication (from the gpg-agent [man page](https://linux.die.net/man/1/gpg-agent) under `--enable-ssh-support`) so if you run `ssh <remote>` on some machine, you are hitting gpg-agent and there isn't, or shouldn't be, an ssh-agent on that machine. So we immediately rule out any mention of ssh-agent as a separate process.

So let me ask you this: suppose I have local/remote1/remote2 and I ssh from local to remote1 and run gpg (or ssh) on remote1. I must have gpg-agent on remote1 for that to work. Now suppose I ssh from remote1 to remote2. Do you think gpg-agent then disappears from remote1? How can that be, since it must still exist if only to continue SSH forwarding from remote1 to remote2. Or suppose not: does it reappear when I return from remote2 to remote1? What if, while connected to remote2, I open up another terminal window on local and ssh to remote1, is gpg-agent there on remote1 or not?

**Comment by jeremyn** [comment]
@jordemort I should clarify that these are rhetorical questions to suggest some of the strangeness that would happen without intermediate gpg-agents. I'm not asking you to actually answer though you can if you want to of course. In fact I feel like we're hitting a wall here unless we want to drag out SSH configs and bashrcs and ps output and screenshots which I don't really want to do.

I would definitely be interested though if you could find some direct authoritative response backing you up, like a comment from a gpg dev on a mailing list saying "of course there are no intermediate gpg-agents, that would be absurd." I could always be wrong.

**Comment by jordemort** [comment]
> So let me ask you this: suppose I have local/remote1/remote2 and I ssh from local to remote1 and run gpg (or ssh) on remote1. I must have gpg-agent on remote1 for that to work.

You are incorrect. You must be able to communicate with a gpg-agent from remote1, but the gpg-agent that you are communicating does not need to be local. You can forward the UNIX socket that gpg-agent is listening on from local to remote1, and sign things on remote1, with no gpg-agent running on remote1 - gpg will communicate over the socket that SSH is forwarding to the gpg-agent running on local.

> Now suppose I ssh from remote1 to remote2. Do you think gpg-agent then disappears from remote1? 

It was never there. Depending on your ssh_config on remote1, you may or may not further forward the gpg-agent socket onward to remote2; if you do, then you should be able to sign things on remote2 as well - again, with no gpg-agent running on remote1 or remote2 - gpg will communicate over the socket forwarded by remote1, which will be forwarded to to the socket forwarded by local, which will talk to the gpg-agent running on local.

> What if, while connected to remote2, I open up another terminal window on local and ssh to remote1, is gpg-agent there on remote1 or not?

Depends on your ssh_config on remote2; if you went from local to remote1 to remote2 and back to remote1, forwarding the socket all along the way, I suspect when you looped back to remote1  you would overwrite the socket you had forwarded from local and break the chain.

It's ssh_config all the way down.

> I could always be wrong.

I am not a GPG developer, but I am a person who has been using a forwarded gpg-agent to sign their commits in their everyday work for five years, and who occasionally has to build containerized CI pipelines that communicate with gpg-agent, and I assure you that you are very persistently and condescendingly wrong. Please take a hike.

**Comment by exilvm** [comment]
> > We only use the extra socket
> 
> I think this is the piece I was missing. I was forwarding the extra socket from my Mac to my Linux host, but I was forwarding it to the regular socket on the Linux host. That is to say, I was forwarding Mac `S.gpg-agent.extra` to Linux `S.gpg-agent`. So there was no functional extra socket on the Linux host.
> 
> I changed my configuration to forward `S.gpg-agent.extra` to both `S.gpg-agent` and `S.gpg-agent.extra` on Linux, and now things just work without any additional hacks - I no longer have to bind `/run/user` or do any special setup in the devcontainer. I can now open up devcontainers and create signed commits without any additional hacks or reconnections.
> 
> For posterity, here is the line I originally had in my `ssh_config`:
> 
> ```
>   RemoteForward /run/user/2001/gnupg/S.gpg-agent /Users/jordan/.gnupg/S.gpg-agent.extra
> ```
> 
> ...and here is the line I added that made things start working:
> 
> ```
>   RemoteForward /run/user/2001/gnupg/S.gpg-agent.extra /Users/jordan/.gnupg/S.gpg-agent.extra
> ```
> 
> Since this now has the "Works on my machine" seal of approval, I'll close this.

Hello, i am facing the same issue but this did not work, i just couldn't forward `S.gpg-agent.extra` to my devcontainer.

**Comment by inkeliz** [comment]
> If I then open a terminal window and start another SSH session to my Ubuntu host, it re-creates the forwarded gpg-agent socket, and then I am able to sign commits again. So I am able to work around it, but I need to open up another terminal window after I open my project in the devcontainer and keep it running as long as I am working, which is somewhat less than idea.

That is exactly what I'm experiencing. I'm using YubiKey and I'm using DevContainer one remote machine. I have the following devcontainer.json:

```
	"build": { "dockerfile": "Dockerfile" },
	"mounts": [
		"source=${localWorkspaceFolder}/.devcontainer/.gnupg,target=/home/vscode/.gnupg,type=bind"
	]
```

_The .gnupg folder is on gitignore, I choose it just to me easier to set RemoteForward._

I have the following ssh-config:

```
Host remote
  HostName remote.my-project.inkeliz
  User remote
  ForwardAgent yes
  RemoteForward /home/remote/project/.devcontainer/.gnupg/S.gpg-agent /Users/inkeliz/.gnupg/S.gpg-agent.extra
  StreamLocalBindUnlink yes
```

If I start the DevContainer in VSCode, and run: `gpg --sign "README.md"` it fails. However, while the DevContainer is running, if I start another SSH connection in my local machine (i.e. `ssh remote` in the terminal), then `gpg --sign "README.md"` inside VSCode: it works.

So, I think my issue is that "S.gpg-agent" is not active when SSH connection is created (when VSCode connects to my server), then when the Docker starts the "S.gpg-agent" is active but the current SSH session don't updates it.

Have any solution?

**Comment by jordemort** [comment]
Hi, I'm tired of this showing up on my list of open issues, when it has been resolved for me for almost 2 years. I am closing this. If you are a maintainer and still feel you have something to need to track regarding this, or if you are a fellow end-user and feel that you have a similar or related problem, please open a new issue and reference this one, rather than re-opening this one. Thanks!
